src/bson.rs

```diff
@@ -1102,6 +1103,21 @@ impl Display for Binary {
 }
 
 impl Binary {
+    /// Creates a [`Binary`] from a base64 string and optional [`BinarySubtype`]. If the
+    /// `subtype` argument is `None`, the [`Binary`] constructed will default to
+    /// [`BinarySubtype::Generic`].
+    pub fn from_base64(
+        input: &str,
+        subtype: impl Into<Option<BinarySubtype>>,
+    ) -> Result<Self, DecodeError> {
```

Comment:
We tend to avoid including types from other crates in our public API as they don't always have the same stability guarantees that we do. Under semantic versioning, any crate with a version that begins with 0 (e.g. 0.13.0, which is the version of the `base64` crate) can make breaking changes at any time -- as an example, `base64` could rename `DecodeError` to `DecodingError`, which would cause compilation failures for any code that relies upon the name being `DecodeError`. Because of this we'll want to avoid returning this type directly.

I looked over the various `Error` types we have defined in the BSON library and I'm not sure any of them are quite what we need for this situation. The closest would be the `SerializationError` under `ser::Error`, but that's more so used in the context of serializing to `Bson`. Alternatively, we could introduce a new `Error` type for this case.

Thoughts on this @patrickfreed @abr-egn @kmahar?

Reply:
Agreed with Patrick on both the immediate thing to do and what we'll want to queue up for 3.0.

Reply:
That sounds good to me. @sanav33 we can walk through what the new error type should look like, I think we'll want to add a `binary` module for it to live in to be consistent with our other error types. Also filed RUST-1406 to clean this up when we do a 3.0 release.